lib/chef/knife/ec_key_import.rb

```diff
@@ -39,17 +40,130 @@ class EcKeyImport < Chef::Knife
+      # backup.
+      #
+      # The org_id is likely different than that stored in the backup.
+      # We query the new org_id based on org_name, caching it to avoid
+      # multiple lookups in a large org.
+      def insert_key_data_for_client(d)
+        ui.msg "Updating key data for client[#{d['name']}]"
+
+        org_id = @org_id_cache.fetch(d['org_name'])
+        if org_id.nil?
+          ui.warn "Could not find organization for client[#{d['name']}], skipping."
+          ui.warn "Organizations much be restored before key data can be imported."
+          return
+        end
+
+        existing_client = db[:clients].where(:org_id => org_id, :name => d['name']).first
+        if existing_client.nil?
+          ui.warn "Did not find existing client record for #{d['name']}, skipping."
+          ui.warn "Client records must be restored before key data can be imported."
+          return
+        end
+
+        new_id = existing_client = new_id
+        Chef::Log.debug("Found client id for #{d['name']}: #{new_id}")
+        upsert_key_record(key_record_for_db(d, new_id))
+      end
+
+      # Insert key data for each user record
+      #
+      # When :skip_id's is set, we are not using the ids from the
+      # backup.  In this case, look up the user id in the users table.
+      #
+      # When :skip_id's is not set, we are using the ids from the
+      # backup. The update_key trigger on the users table should
+      # ensure that the user ids have already been replaced and should
+      # match what we have in the backup.
+      def insert_key_data_for_user(d)
+        if d['name'] == 'pivotal' && config[:skip_pivotal]
+          ui.warn "Skipping pivotal user."
+          return
+        end
+        ui.msg "Updating key data for user[#{d['name']}]"
+        new_id = if config[:skip_ids]
+                   db[:users].where(:username => d['name']).first[:id]
+                 else
+                   d['id']
+                 end
+        Chef::Log.debug("Found user id for #{d['name']}: #{new_id}")
+        upsert_key_record(key_record_for_db(d, new_id))
+      end
+
+      def upsert_key_record(r)
+        key_records_to_update = db[:keys].where(:key_name => r[:key_name], :id => r[:id])
+        case key_records_to_update.count
+        when 0
+          Chef::Log.debug("No existing records found for #{r[:key_name]}, #{r[:id]}")
+          db[:keys].insert(r)
+        when 1
+          Chef::Log.debug("1 record found for #{r[:key_name]}, #{r[:id]}")
+          key_records_to_update.update(r)
+        else
+          ui.warn "Found too many records for actor id #{r[:id]} key #{d[:key_name]}"
+          return
+        end
+      end
+
+      def key_record_for_db(d, new_id=nil)
+        {
+          :id => new_id || d['id'],
+          :key_name => d['key_name'],
+          :public_key => d['public_key'],
+          :key_version => d['key_version'],
+          :created_at => Time.now,
```

Comment:
We could.  We'd either have to change the keys_by_name view or construct a new query.  I didn't worry about this here since nearly all the objects we "restore" won't have created_at preserved since we create them via the API.